lib/components/rails/component.rb

```diff
@@ -26,7 +27,11 @@ def initialize(object, view, action, attributes = {})
         @attributes = attributes
         @action = action
 
-        @block_content = attributes[:block].call if attributes[:block]
+        if attributes[:block]
+          value = nil
+          buffer = with_output_buffer { value = attributes[:block].call }
+          @block_content = buffer.presence || value
+        end
```

Comment:
@benhutton View the source of this method https://apidock.com/rails/ActionView/Helpers/CaptureHelper/capture to get an idea where I got this from.

Reply:
@joeosburn-desiringgod cool. The `with_output_buffer` thing is essentially what I would expect. Can you explain what's going on with the `value` instance var?

That is, why can't we just do

```
if attributes[:block]
  @block_content = with_output_buffer { attributes[:block].call }
end
```

Reply:
@benhutton Yep...

So, just to be clear, `#with_output_buffer` captures a block that is generating output that is going to the output buffer (in addition to returning the string.) We can't control the fact that a block is outputting to the output buffer. The `do..end` erb block, for example, is in a template, and it's going to output to the output buffer, and that's beyond our control, because the block is running in the template context.

`with_output_buffer` is tricky, because it swaps the output buffers before and after the block runs, so that we can grab that output buffer content and treat it like a normal string return.

However, there may be some situations where the block we have doesn't actually output to the output buffer and we still need to handle it in those situations. That's what the `value = ` business is for.  The only situation I know of it _actually_ occurring in our codebase is in the spec, here: https://github.com/desiringgod/components-rails/blob/4940260e799d47c09846591e645d92bd462cb7c5/spec/features/rendering_spec.rb#L103, so in one sense, it's only useful for us. However, you'll note that this same bit of logic is also present in the capture helper, so I think it's useful to retain it. It may be that other template formats (haml, simple, etc.) work differently, or this simple allows the content block to work outside of template contexts.

The last point of note here is that I'm not keeping the `String` checking logic that was in the capture helper. I'm not sure what the significance of that is, and when/if we start to run into issues with non-string values, we can look into it then.

Reply:
@joeosburn-desiringgod okay, I think I'm understanding... you're saying that sometimes `with_output_buffer` isn't going to actually return something, for whatever reason, but we can still get a value by assigning to `value`

Reply:
@benhutton Yep.